crates/sui-indexer-alt/src/handlers/sum_obj_types.rs

```diff
@@ -147,40 +162,74 @@ impl Handler for SumObjTypes {
+            let object_version = update.object_version;
+            match (&update.kind, &update.value) {
+                (UpdateKind::Insert, Some(value)) => inserts.push(value.clone()),
+                (UpdateKind::Update, Some(value)) => updates.push(value.clone()),
+                (UpdateKind::Delete, None) => deletes.push(update.object_id.to_vec()),
+
+                (UpdateKind::Insert | UpdateKind::Update, None) => {
+                    bail!(
+                        "Missing contents for object {} version {}",
+                        object_id.to_canonical_display(/* with_prefix */ true),
+                        object_version,
+                    );
+                }
+
+                (UpdateKind::Delete, Some(_)) => {
+                    bail!(
+                        "Unexpected contents for deleted object {} version {}",
+                        object_id.to_canonical_display(/* with_prefix */ true),
+                        object_version,
+                    );
+                }
             }
         }
 
-        let update_chunks = updates.chunks(UPDATE_CHUNK_ROWS).map(Either::Left);
+        let insert_chunks = inserts
+            .chunks(UPDATE_CHUNK_ROWS)
+            .map(|i| Either::Left(Either::Left(i)));
+
+        let update_chunks = updates
+            .chunks(UPDATE_CHUNK_ROWS)
+            .map(|u| Either::Left(Either::Right(u)));
+
         let delete_chunks = deletes.chunks(DELETE_CHUNK_ROWS).map(Either::Right);
 
-        let futures = update_chunks.chain(delete_chunks).map(|chunk| match chunk {
-            Either::Left(update) => Either::Left(
-                diesel::insert_into(sum_obj_types::table)
-                    .values(update)
-                    .on_conflict(sum_obj_types::object_id)
-                    .do_update()
-                    .set((
-                        sum_obj_types::object_version.eq(excluded(sum_obj_types::object_version)),
-                        sum_obj_types::owner_kind.eq(excluded(sum_obj_types::owner_kind)),
-                        sum_obj_types::owner_id.eq(excluded(sum_obj_types::owner_id)),
-                    ))
-                    .execute(conn),
-            ),
-
-            Either::Right(delete) => Either::Right(
-                diesel::delete(sum_obj_types::table)
-                    .filter(sum_obj_types::object_id.eq_any(delete.iter().cloned()))
-                    .execute(conn),
-            ),
-        });
+        let futures = insert_chunks
+            .chain(update_chunks)
+            .chain(delete_chunks)
+            .map(|chunk| match chunk {
+                Either::Left(Either::Left(insert)) => Either::Left(Either::Left(
+                    diesel::insert_into(sum_obj_types::table)
+                        .values(insert)
+                        .execute(conn),
+                )),
+                Either::Left(Either::Right(update)) => Either::Left(Either::Right(
+                    update_from(sum_obj_types::table)
```

Comment:
Yeah, the complexity in `diesel-update-from` is not great. The rationale behind isolating it in its own crate is that it is unlikely to change, and we should treat it as if it were part of diesel, so maintenance should be minimal but I would be lying if I said that there would be none (the fact that it requires opting in to breaking changes alone is a sign that there may be some).

To me this is the argument for not using `diesel`, though. If it does not support all the features we need, and we do not feel able to add and maintain those features ourselves, it is going to bias us towards doing things in a sub-optimal way to fit into the feature set that diesel offers conveniently, which we know we cannot afford to do and this is not the first time this has bitten us:

- Our inserts and updates bind every value rather than UNNEST arrays of values.
- We can't write comparisons over tuples, which would help postgres plan those queries more efficiently.
- We can't use CTEs
- ...I am sure there are other things that I can't remember, and this is not to mention that the errors we get out of diesel when we make a mistake are horrific.

In each of these occasions, we either put up with the inefficiency or drop into raw queries where we don't get support for detecting schema changes, dealing with type marshaling etc, and the story would be similar here -- it is possible to use raw queries but it defeats the point of the abstraction if we lose its safety features when the query gets more complex.

Reply:
Remembered another one -- diesel's query types don't support `Clone` so we needed to bend over backwards in GraphQL to do stuff like "explain a query first, and then run it", or "print out the query".